src/modules/fancyzones/FancyZonesLib/Drawing.cpp

```diff
@@ -0,0 +1,240 @@
+#include "pch.h"
+#include "Drawing.h"
+
+#include <common/logger/logger.h>
+
+D2D1_COLOR_F Drawing::ConvertColor(COLORREF color)
+{
+    return D2D1::ColorF(GetRValue(color) / 255.f,
+                        GetGValue(color) / 255.f,
+                        GetBValue(color) / 255.f,
+                        1.f);
+}
+
+ID2D1Factory* Drawing::GetD2DFactory()
+{
+    static auto pD2DFactory = [] {
+        ID2D1Factory* res = nullptr;
+        D2D1CreateFactory(D2D1_FACTORY_TYPE_MULTI_THREADED, &res);
+        return res;
+    }();
+    return pD2DFactory;
+}
+
+IDWriteFactory* Drawing::GetWriteFactory()
+{
+    static auto pDWriteFactory = [] {
+        IUnknown* res = nullptr;
+        DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &res);
+        return reinterpret_cast<IDWriteFactory*>(res);
+    }();
+    return pDWriteFactory;
+}
+
+IWICImagingFactory2* Drawing::GetImageFactory()
+{
+    static auto pImageFactory = [] {
+        PVOID res = nullptr;
+        CoCreateInstance(CLSID_WICImagingFactory2, nullptr, CLSCTX_INPROC_SERVER | CLSCTX_LOCAL_SERVER, __uuidof(IWICImagingFactory2), &res);
+        return reinterpret_cast<IWICImagingFactory2*>(res);
+    }();
+    return pImageFactory;
+}
+
+Drawing::Drawing()
+{
+    m_window = nullptr;
+    m_renderTarget = nullptr;
+}
+
+void Drawing::Init(HWND window)
+{
+    m_window = window;
+
+    // Obtain the size of the drawing area.
+    if (!GetClientRect(window, m_renderRect.get()))
+    {
+        Logger::error("couldn't initialize Drawing: GetClientRect failed");
+        return;
+    }
+
+    // Create a Direct2D render target
+    // We should always use the DPI value of 96 since we're running in DPI aware mode
+    auto renderTargetProperties = D2D1::RenderTargetProperties(
+        D2D1_RENDER_TARGET_TYPE_DEFAULT,
+        D2D1::PixelFormat(DXGI_FORMAT_UNKNOWN, D2D1_ALPHA_MODE_PREMULTIPLIED),
+        96.f,
+        96.f);
+
+    auto renderTargetSize = D2D1::SizeU(m_renderRect.width(), m_renderRect.height());
+    auto hwndRenderTargetProperties = D2D1::HwndRenderTargetProperties(window, renderTargetSize);
+
+    winrt::com_ptr<ID2D1HwndRenderTarget> renderTarget = nullptr;
+    auto hr = GetD2DFactory()->CreateHwndRenderTarget(renderTargetProperties, hwndRenderTargetProperties, renderTarget.put());
+    if (!SUCCEEDED(hr))
+    {
+        Logger::error("couldn't initialize Drawing: CreateHwndRenderTarget failed with {}", hr);
+        return;
+    }
+
+    m_renderTarget = renderTarget;
+}
+
+Drawing::operator bool() const
+{
+    return m_renderTarget != nullptr;
+}
+
+void Drawing::BeginDraw(const D2D1_COLOR_F& backColor)
+{
+    m_renderTarget->BeginDraw();
```

Comment:
The same thing with an initialization, it's unsafe to leave it like this. Let's add checks in public methods.

Reply:
Here the checks are in lines 189 and 241. Same thing.